File: linear-algebra/coordinate-transformation/xy2sn.m
% Fri Jul  4 19:01:06 WIB 2014
% Karl Kastner, Berlin
%
%% convert cartesian to streamwise coordiantes
%
function [S N] = xy2sn2(cX,cY,cS,X,Y)

	% for each point in X, find closest centreline point and distance to it
	S = zeros(size(X));
	N = zeros(size(X));
	% find nearest neighbours on the centreline for each input point
	nn = knnsearch([cX cY],[X Y],'K',1);

	for idx=1:length(X)
		% get the index of closest centreline point
		mx1 = nn(idx);
	
		% determine second neighbour
		% TODO this should not be second closest but
		%      the neighbour that makes the transformation function convex,
		%      e.g. [X Y] = p [cX1 + (1-p) cX2, 0 <= p <= 1
		if (1 == mx1)
			mx2 = mx1+1;
		elseif (mx1 == length(cX))
			mx2 = mx1-1;
		else
			d2left  = Geometry.distance2([X(idx) Y(idx)],[cX(mx1-1) cY(mx1-1)]);
			d2right = Geometry.distance2([X(idx) Y(idx)],[cX(mx1+1) cY(mx1+1)]);
			if (d2left < d2right)
				mx2 = mx1-1;
			else
				mx2 = mx1+1;
			end
		end
		% corner vertices of the triangle
		A = [cX(mx1); cY(mx1)];
		B = [cX(mx2); cY(mx2)];
		C = [X(idx); Y(idx)];

		% translace coordinates such that A is the origin
		B = B-A; 
		C = C-A;
		% A := 0
		% rotate coordinates, such that AB = (B-0) becomes oriented along the x-axis
		% C := R^-1 C
		hyp = sqrt(B'*B);
		sina = B(2)/hyp;
		cosa = B(1)/hyp;
		Ri = [ cosa sina
	              -sina cosa];
		C = Ri*C;
		% now Cx is the dS and Cy is N
		if (cS(mx1) < cS(mx2))
			S(idx) = cS(mx1) + C(1);
			N(idx) = C(2);
		else
			S(idx) = cS(mx1) - C(1);
			N(idx) = -C(2);
		end
%		S_(idx,:) = [cS(mx1) S(idx) cS(mx2)]-cS(mx1);
%		catch e
%			e
%			S(idx) = NaN;
%			N(idx) = NaN;
%		end
	end % for idx
%	close all
%	plot([S_]);
%	pause

end % xy2sn

File: linear-algebra/coordinate-transformation/XY2sn.java
// Wed Jun 25 14:04:35 WIB 2014
// Karl Kastner, Berlin
// convert cartesian to streamwise coordinates and back

import java.lang.*;

class XY2sn
{
	// variables for data exchange with matlab
	public double [] S;
	public double [] N;
	public int [][] Nkey;

	// default constructor
	public XY2sn() {}

	// final causes also problems in matlab
	public void xy2sn_(final double [] cX,
			   final double [] cY,
		           final double [] cS,
			   final int    [] cSeg,
                           final double [] X,
                           final double [] Y,
                           double [] S,
                           double [] N) throws Exception
	{
//System.out.println(array.getClass());
		// aux variables
		// find nearest neighbour
		Qtree2 qtree      = new Qtree2(cX,cY);
		int [][] Nkey    = new int[X.length][1];
		double [][] Dmin = new double[X.length][1]; 
		qtree.nearest_neighbour(X, Y, Nkey, Dmin);

		// for each point in X, find closest centreline point and distance to it
		for (int idx=0; idx<X.length; idx++)
		{
			// get closest centre line point
			int mx1 = Nkey[idx][0]-1;
			int mx2;
			// determine second smallest element
			// take left neighbour if it exists
			// no right neighbour exists
			// or distance to left neighbour is smaller
			if (-1 == mx1)
			{
				// no neighbour found likely reason is a NaN coordinate of the current point
				S[idx] = Double.NaN;
				N[idx] = Double.NaN;
				continue;
			} else if (0 == mx1 || cSeg[mx1-1] != cSeg[mx1])
			{
				// first element or first element in this segment
				if (cX.length-1 != mx1 && cSeg[mx1] == cSeg[mx1+1])
				{
					// mx1 is leftmost element,
					// and there is a right neighbour, choose it
					mx2 = mx1+1;
				} else {
					// this segment is of unit length, 
					// projection impossible
					S[idx] = Double.NaN;
					N[idx] = Double.NaN;
					continue;
				}
			} else if(cX.length-1 == mx1 || cSeg[mx1] != cSeg[mx1+1])
			{
				// last element or last element in this segment
				if (0 != mx1 && cSeg[mx1-1] == cSeg[mx1])
				{	
					// mx1 is rightmost element,
					// and there is a left neighbour, choose it
					mx2 = mx1-1;
				} else {
					// this segment is of unit length, 
					// projection impossible
					S[idx] = Double.NaN;
					N[idx] = Double.NaN;
					continue;
				}
			} else {
			// TODO this should not be the second closest but the second neighbour leading to a convex S
				double d2left  = dist2(X[idx],Y[idx],cX[mx1-1],cY[mx1-1]);
				double d2right = dist2(X[idx],Y[idx],cX[mx1+1],cY[mx1+1]);

				if (d2left < d2right)
				{
					mx2 = mx1-1;
				} else {
					mx2 = mx1+1;
				}
			}
			// corner vertices of the triangle
			double Ax = cX[mx1];
			double Ay = cY[mx1];
			double Bx = cX[mx2];
			double By = cY[mx2];
			double Cx = X[idx];
			double Cy = Y[idx];
			// translate 
			// make A the local coodinate origin
			Bx = Bx-Ax;
			By = By-Ay;
			Cx = Cx-Ax;
			Cy = Cy-Ay;

			// rotate
			double hyp = Math.hypot(Bx,By); 
			double sina = (1.0/hyp)*By;
			double cosa = (1.0/hyp)*Bx;
			double Cx_ =  cosa*Cx + sina*Cy;
			double Cy_ = -sina*Cx + cosa*Cy;

			if (cS[mx1] < cS[mx2])
			{
				S[idx] = cS[mx1]+Cx_;
				N[idx] = Cy_;
			} else {
				S[idx] = cS[mx1]-Cx_;
				N[idx] = -Cy_;
			}
	} // for idx
			// store values for matlab communication
			this.S = S;
			this.N = N;
			this.Nkey = Nkey;
	} // function xy2sn

	private final static double dist2( final double ax, final double ay, final double bx, final double by)
	{
		return (bx-ax)*(bx-ax) + (by-ay)*(by-ay);
	}
			
	private final static double dot(final double ax, final double ay, final double bx, final double by)	
	{
		return (ax*bx)+(ay*by);
	} // function dot
} // class xy2sn

